File: Python/AdapterPattern.py
# Adaptee - EuropeanSocket
class EuropeanSocket:
    def plug(self):
        return "Plugged into European socket."

# Target Interface - SocketAdapter
class SocketAdapter:
    def connect(self):
        pass

# Adapter - EuropeanToUSAdapter
class EuropeanToUSAdapter(SocketAdapter):
    def __init__(self,socket):
        self.socket = socket

    def connect(self):
        return self.socket.plug()

# Client code
def charge_laptop(socket_adapter):
    return socket_adapter.connect()

# Usage
european_socket = EuropeanSocket()
adapter = EuropeanToUSAdapter(european_socket)
print(charge_laptop(adapter))

File: Javascript/AdapterPattern.js
// Adaptee - EuropeanSocket
class EuropeanSocket {
  plug() {
    return "Plugged into European socket";
  }
}

// Target Interface - SocketAdapter
class SocketAdapter {
  connect() {
    throw new Error("Method not implemented.");
  }
}

// Adapter - EuropeanToUSAdapter
class EuropeanToUSAdapter extends SocketAdapter {
  constructor(socket) {
    super();
    this.socket = socket;
  }

  connect() {
    return this.socket.plug();
  }
}

// Client Code
function chargeLaptop(socketAdapter) {
  return socketAdapter.connect();
}

// Usage
const europeanSocket = new EuropeanSocket();
const adapter = new EuropeanToUSAdapter(europeanSocket);
console.log(chargeLaptop(adapter));
